[No QA] Add test for matching Translations keys

parasharrajat opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Added Unit test that matches the keys from each Language file to the Main EN.js translations. If any key is missing, the test will fail. You will see a log stating the path of the key.

### Fixed Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing -->
$ #4082

parasharrajat committed:
add test for Translations keys match

parasharrajat commented:
@iwiznia this PR is itself a test the PR. Check this  https://github.com/Expensify/Expensify.cash/pull/4090/checks?check_run_id=3080986765#step:6:722.

parasharrajat committed:
added translations

parasharrajat commented:
Please review the translations before merging. I don't speak Spanish but took help from google.

iwiznia commented on tests/unit/TranslateTest.js:
```diff
@@ -52,3 +54,38 @@ describe('translate', () => {
+        return true;
+    }
+    it('Does each locale has all the keys', () => {
+        const excludeLanguages = ['en', 'es-ES'];
+        const languages = _.without(_.keys(originalTranslations.default), ...excludeLanguages);
```
Can we make this smarter and instead of having to add `es-ES` here make it skip any file that has 2 components? ie: pick up `es` but not `es-ES`

parasharrajat replied:
That was not the original Idea. After Sometime, we can decide to add more Translations. and Two component locale would be completely valid in that case. So to skip regardless of specification will lead to code change. So I think its better, not to make it smarter. It is sufficient to fulfill its purpose.

iwiznia commented on tests/unit/TranslateTest.js:
```diff
@@ -52,3 +54,38 @@ describe('translate', () => {
 });
+
+describe('Translation Keys', () => {
+    let activeLanguage;
```
Why is this "global"? Can't we pass it as a regular argument to `matchKeys`?

parasharrajat replied:
This is convenient.

iwiznia commented on tests/unit/TranslateTest.js:
```diff
@@ -52,3 +54,38 @@ describe('translate', () => {
+        path += key ? `${key}.` : '';
+        const pathLevel = path;
+        if (key && !_.has(target, key)) {
+            console.debug(`🏹 ${path.slice(0, -1)} is missing from ${activeLanguage}.js`);
```
Instead of a debug log line, can we make it so that the test itself fails? Ideally it would show one failed test per missing key if that's possible, but if that's hard we can have one test fail that prints all missing keys.

parasharrajat replied:
Test will fail, if key don't match.  This log is just to show what is missing. 

> Ideally it would show one failed test per missing key if that's possible,

I don't think its a good idea to create multiple tests that basically test the same thing. 

> if that's hard we can have one test fail that prints all missing keys.

This can be done. We can log all the missing keys.

iwiznia replied:
ok cool, my main point is that it's better to have the test failure show what keys failed instead of having to look at the logs.

parasharrajat replied:
I tried to find way to add the missing key as failure message itself but I don't think its possible AFAIK. So we have to stick with logs.

iwiznia replied:
hmmmm, I think that if you change the expect from `toBeTruthy` to something like `arrayContaining` would make jest to print the diff of the arrays and thus the missing keys?

iwiznia reviewed (changes requested):
Sorry that I did not realize about this when creating the issue, but can we also add the opposite check? Basically to ensure we don't have tons of unused keys in the non base english files, we would also alert for any key present in non base english files that are not present on the english files.

parasharrajat commented:
> Basically to ensure we don't have tons of unused keys in the non base english files, we would also alert for any key present in non base english files that are not present on the english files.

Hmm, let me think about it.

iwiznia commented:
Any update here @parasharrajat? I see we are missing translations in several places and this would prevent all those errors.

parasharrajat commented:
@iwiznia I am thinking to rewrite the tests with a simpler and robust implementation. I will post multiple test clauses for each language file...etc. Maybe I'll finish it today.

parasharrajat committed:
fix: translation tests

parasharrajat committed:
fix translation files

parasharrajat committed:
Pull latest changes

parasharrajat committed:
added missing key

parasharrajat commented:
@iwiznia It's ready. `Expect.arrayContaining` is not useful as it does not give the required output and does not tell us which keys are missing. WRT, it outputs huge string on the logs which basically make it non-readable. So I have created our on Logs using `console.log`.

iwiznia commented on tests/unit/TranslateTest.js:
```diff
@@ -52,3 +54,42 @@ describe('translate', () => {
+    const parentLanguage = originalTranslations.default.en;
+    const parentLanguageKeys = traverseKeyPath(parentLanguage);
+
+    _.every(languages, (ln) => {
```
You are not really using every here, you are just doing something on each element of the array, so this is probably better:
```suggestion
    _.each(languages, (ln) => {
```

parasharrajat replied:
Nice catch

iwiznia commented on tests/unit/TranslateTest.js:
```diff
@@ -52,3 +54,42 @@ describe('translate', () => {
+    }
+    const excludeLanguages = ['en', 'es-ES'];
+    const languages = _.without(_.keys(originalTranslations.default), ...excludeLanguages);
+    const parentLanguage = originalTranslations.default.en;
```
Kind of an odd name, I suggest:
```suggestion
    const mainLanguage = originalTranslations.default.en;
```

iwiznia reviewed (commented):
Looks great, but you got conflicts. Can you:
1. Resolve the conflicts
2. Make some tests fail in one of your commits, so I can look at the output of the failing tests in the GH actions
Please?

parasharrajat committed:
refactor tests

parasharrajat committed:
pulled latest changes and Show test failure

parasharrajat commented:
@iwiznia I have made few changes to es.js which will result in a test failure. Please check https://github.com/Expensify/App/pull/4090/checks?check_run_id=3192256522#step:6:595.

iwiznia commented:
Ah damn, that's not that great as the exact failures are mixed in with other non useful logs. Making some suggestions in the PR to see if we can make it better.

iwiznia commented on tests/unit/TranslateTest.js:
```diff
@@ -52,3 +54,42 @@ describe('translate', () => {
+            if (hasAllKeys.length) {
+                console.debug(`🏹 [ ${hasAllKeys.join(', ')} ] are missing from ${ln}.js`);
+            }
+            expect(hasAllKeys.length).toBe(0);
```
So, I was thinking, maybe we could do this and get a better error message?
```suggestion
            expect(hasAllKeys).toBeEqual([]);
```
I think this will show in the test result exactly which keys are missing, no?
(same applies to the one below)

parasharrajat replied:
It does not show the keys that are not correct. It just shows Expected and produced output. So all the correct and wrong keys would be shown as a big string which is like 50 lines long.

parasharrajat replied:
I tested with `expect.objectContaining and expect.arrayContaining`.

iwiznia replied:
hmmmm, can we change the code to do `_.without(mainLagunageKeys, ...languageKeys)`? That would return only the missing keys, right?

parasharrajat replied:
It currently only give the missing keys. Again `_.without(mainLagunageKeys, ...languageKeys)` will go in `console.log`. @iwiznia

iwiznia replied:
The console log is not good enough as it is super hard to find the right message. In any case I don't think you understood what I meant, because I just tried and it works. Here's the test code:
```
        it(`Does ${ln} locale has all the keys`, () => {
            const hasAllKeys = _.without(mainLanguageKeys, ...languageKeys);
            expect(hasAllKeys).toEqual([]);
        });

        it(`Does ${ln} locale has unused keys`, () => {
            const hasAllKeys = _.without(languageKeys, ...mainLanguageKeys);
            expect(hasAllKeys).toEqual([]);
        });
```
And this is the output:
```
 ● Translation Keys › Does es locale has all the keys

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 4

    - Array []
    + Array [
    +   "preferencesPage.mostRecent",
    +   "passwordForm.error.incorrectLoginOrPassword",
    + ]

      79 |         it(`Does ${ln} locale has all the keys`, () => {
      80 |             const hasAllKeys = _.without(mainLanguageKeys, ...languageKeys);
    > 81 |             expect(hasAllKeys).toEqual([]);
         |                                ^
      82 |         });
      83 |
      84 |         it(`Does ${ln} locale has unused keys`, () => {

      at Object.<anonymous> (tests/unit/TranslateTest.js:81:32)

  ● Translation Keys › Does es locale has unused keys

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 4

    - Array []
    + Array [
    +   "initialSettingsPage.appDownloadLinks.window.label",
    +   "initialSettingsPage.appDownloadLinks.Rajat.label",
    + ]

      84 |         it(`Does ${ln} locale has unused keys`, () => {
      85 |             const hasAllKeys = _.without(languageKeys, ...mainLanguageKeys);
    > 86 |             expect(hasAllKeys).toEqual([]);
         |                                ^
      87 |         });
      88 |     });
      89 | });

      at Object.<anonymous> (tests/unit/TranslateTest.js:86:32)

```

parasharrajat replied:
oh Yeah. This way I got it now.

parasharrajat replied:
@iwiznia  Check this out https://github.com/Expensify/App/actions/runs/1079637864. I added annotations.
![image](https://user-images.githubusercontent.com/24370807/127545308-f5290a18-16b7-4567-8555-1bb20fc5ab34.png)

parasharrajat committed:
fix: tests

parasharrajat commented:
@iwiznia Thanks for the hint. I have updated the PR with the following changes:
1. Your suggestion of `toBeEqual([])`.
2. Added Annotations to quickly see what is the cause of faliure.
3. Fixed all the translations and Finalized the PR.

It's now ready to be merged.

iwiznia reviewed (approved):
Ah lovely!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1083388541) to staging         in version: 1.0.81-5🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1103640379) to production         in version: 1.0.82-7🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
